Prepend program name to args after merge

am11 opened the pull request:
Fixes #942

fabioDMFerreira reviewed (changes requested):
Thanks for this @am11.

We need to handle the args flag in every command that needs to build an image. To avoid setting the program name as the first argument in each command handler we can do it on the struct responsible for merging the build image configuration (https://github.com/nanovms/ops/blob/master/cmd/flags_build_image.go#L65).

If we're using a package configuration to build an image, ops have to merge packages arguments (https://github.com/nanovms/ops/blob/master/cmd/flags_pkg.go#L51). Wondering whether this change will affect the behaviour of `ops image create --package <package> ...` or `ops load <package> ...`

am11 commented:
@fabioDMFerreira, I've addressed the feedback, this is more of a correctness fix. Any suggestions where to put unit test for this so it don't regress?

fabioDMFerreira commented:
You could add a `Program` here and check whether the final arguments are the expected ones
https://github.com/nanovms/ops/blob/master/cmd/flags_build_image_test.go#L53.

I also think it worth to add some dedicated test to check arguments here
https://github.com/nanovms/ops/blob/master/cmd/flags_pkg_test.go. See next an example of a package configuration.
```
{
  "Program": "mysql_5.7.29/mysqld",
  "Version": "5.7.29",
  "RunConfig": {
    "Memory": "2G"
  },
  "BaseVolumeSz": "200m",
  "Args":["mysqld", "--user=root", "--explicit_defaults_for_timestamp"]
}
```

fabioDMFerreira commented on cmd/flags_build_image.go:
```diff
@@ -79,6 +79,8 @@ func (flags *BuildImageCommandFlags) MergeToConfig(c *types.Config) (err error)
 		c.Args = flags.CmdArgs
 	}
 
+	c.Args = append([]string{c.Program}, c.Args...)
```
Worth to check if c.Program isn't empty as the program may come from a package.

am11 commented:
It looks like ops-pkg command with current release (0.1.19) does not respect `Args` with `-c config.json`:

```sh
$ cat > test1.js <<<'console.log("args: ", process.argv)'
$ cat > config.json <<<'{ "Args":["one","two","three"] }'

$ ops pkg load node_v11.5.0 -p 8083 -f -n -a test1.js -c config.json
booting /home/am11/.ops/images/test1.js.img ...
You specified hardware acceleration, but it is not supported
Are you running inside a vm? If so disable accel with --accel=false

en1: assigned 10.0.2.15
args:  [ '/node_v11.5.0/node', '/test1.js' ]
exit status 1
```

on the other hand, the command-line args work:

```sh
$ ops pkg load node_v11.5.0 -p 8083 -f -n -a test1.js -a one -a two -a three
Downloading.. https://storage.googleapis.com/nanos/release/nightly/nanos-nightly-linux.tar.gz
 100% |████████████████████████████████████████|  [0s:0s]
booting /home/am11/.ops/images/test1.js.img ...
You specified hardware acceleration, but it is not supported
Are you running inside a vm? If so disable accel with --accel=false

en1: assigned 10.0.2.15
args:  [ '/node_v11.5.0/node', '/test1.js', 'one', 'two', 'three' ]
exit status 1
```

is this by design?

fabioDMFerreira commented:
> It looks like ops-pkg command with current release (0.1.19) does not respect `Args` with `-c config.json`:
> 
> ```shell
> $ cat > test1.js <<<'console.log("args: ", process.argv)'
> $ cat > config.json <<<'{ "Args":["one","two","three"] }'
> 
> $ ops pkg load node_v11.5.0 -p 8083 -f -n -a test1.js -c config.json
> booting /home/am11/.ops/images/test1.js.img ...
> You specified hardware acceleration, but it is not supported
> Are you running inside a vm? If so disable accel with --accel=false
> 
> en1: assigned 10.0.2.15
> args:  [ '/node_v11.5.0/node', '/test1.js' ]
> exit status 1
> ```
> 
> on the other hand, the command-line args work:
> 
> ```shell
> $ ops pkg load node_v11.5.0 -p 8083 -f -n -a test1.js -a one -a two -a three
> Downloading.. https://storage.googleapis.com/nanos/release/nightly/nanos-nightly-linux.tar.gz
>  100% |████████████████████████████████████████|  [0s:0s]
> booting /home/am11/.ops/images/test1.js.img ...
> You specified hardware acceleration, but it is not supported
> Are you running inside a vm? If so disable accel with --accel=false
> 
> en1: assigned 10.0.2.15
> args:  [ '/node_v11.5.0/node', '/test1.js', 'one', 'two', 'three' ]
> exit status 1
> ```
> 
> is this by design?

Not sure how this should work if we have arguments on command flags and on the configuration file. My first impulse is to say those arguments should be concatenated. However, other configuration properties are overridden if the arguments are specified on command flags.

Any thoughts about this? @eyberg @am11

am11 commented:
@fabioDMFerreira, the problem described in https://github.com/nanovms/ops/pull/947#issuecomment-795354646 is that `ops pkg`, unlike other commands, disregards `Args` in `config.json` when building the image. Should it be changed? This PR hasn't touched ops-pkg because of this.

Regarding override logic (which applies to all commands), I think current approach is a good tradeoff, i.e. command line arguments (`-a foo`) will take precedence over `Args` in config.json as a whole. Ideally we can take the "union" of arguments, but that would require us to implement all kinds of argument parsing behaviors. e.g.  `ops run myApp -a -p:MyParameter="foo ; \"bar\", 'baz'"`.. We can open a separate issue to discuss it in more details..

fabioDMFerreira commented:
This change might affect `ops pkg`. If it keeps running smoothly PR is ok

am11 commented:
As far as I can tell it has no effect on ops-pkg whatsoever.. if you have a repro please let me know.

am11 commented:
Opened #952 for `ops pkg -c confgWithArgs.json` vs. `ops pkg -a foo` behavior mismatch.
